src/csharp/Grpc.Tools.Tests/scripts/fakeprotoc.py

```diff
@@ -0,0 +1,327 @@
+            write_debug("    line: "+line)
+            protoc_args.append(line)
+
+def parse_protoc_arguments():
+    """
+    Parse the protoc arguments that are in protoc_args
+    """
+    global protoc_args
+    global dependencyfile
+    global grpcout
+    global protofile
+
+    write_debug("\nparse_protoc_arguments")
+    for arg in protoc_args:
+        if dbg:
+            write_debug("Parsing: "+arg)
+
+        # All arguments containing file or directory paths are
+        # normalised by converting to relative paths to the
+        # project directory, and all '\' and changed to '/'
+        if arg.startswith("--"):
+            (name, value) = arg.split("=",1)
+
+            if name == "--dependency_out":
+                value = relative_to_project(value)
+                dependencyfile = value
+            elif name == "--grpc_out":
+                value = relative_to_project(value)
+                grpcout = value
+            elif name in [ "--grpc_out", "--proto_path", "--csharp_out" ]:
+                value = relative_to_project(value)
+
+            add_protoc_arg_to_dict(name, value)
+
+        elif arg.startswith("#"):
+            pass # ignore
+        else:
+            # proto file name
+            protofile = relative_to_project(arg)
+            add_protoc_arg_to_dict("protofile", protofile)
+
+def add_protoc_arg_to_dict(name, value):
+    """
+    Add the arguments with name/value to protoc_args_dict
+
+    protoc_args_dict is later used from writing out the JSON
+    results file
+    """
+    global protoc_args_dict
+    if name in protoc_args_dict:
+        values = protoc_args_dict[name]
+        values.append(value)
+    else:
+        protoc_args_dict[name] = [ value ]
+
+def relative_to_project(file):
+    """ Convert a file path to one relative to the project directory """
+    return normalise_slashes(os.path.relpath(os.path.abspath(file), FAKEPROTOC_PROJECTDIR))
+
+def normalise_slashes(path):
+    """ Change all backslashes to forward slashes """
+    return path.replace("\\","/")
+
+def write_results_json(pf):
+    """ Write out the results JSON file """
+    global protoc_args_dict
+    global results_json
+    global FAKEPROTOC_PROJECTDIR
+
+    # Read existing json.
+    # Since protoc may be called more than once each build/test if there is
+    # more than one protoc file, we read the existing data to add to it.
+    fname = os.path.abspath(FAKEPROTOC_PROJECTDIR+"/log/"+FAKEPROTOC_TESTID+".json")
+    if os.path.isfile(fname):
+        results_json = json.load(open(fname,"r"))
+        protoc_files_dict = results_json.get("Files")
+    else:
+        results_json = {}
+        protoc_files_dict = {}
+        results_json["Files"] = protoc_files_dict
+    
+    protofiles = protoc_args_dict.get("protofile")
+    if protofiles is None:
+        key = "NONE"
+    else:
+        key = protofiles[0]
+    results_json["Metadata"] = { "timestamp": str(datetime.datetime.now()) }
```

Comment:
Metadata is just for test developers - e.g. adding a timestamp so that can see when data was generated.
It may become useful later when testing incremental builds - e.g. checking timestamp to check that it didn't run twice